### PR TITLE
perf: optimize circom proof generation

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR optimizes circom proof generation by avoiding parsing `zkey` and `wtns` files.

Related: #460

### CODE REVIEW

**Comment by TomTaehoonKim** [comment]
In 134c146, "not moved" -> "not be moved" would be better.

**Comment by TomTaehoonKim** [comment]
In 63d388a, there is a typo "pref" -> "perf".

File: tachyon/base/buffer/read_only_buffer.h [hunk 1]
```diff
@@ -133,6 +134,22 @@ class TACHYON_EXPORT ReadOnlyBuffer {
...
+  template <typename T>
+  [[nodiscard]] bool ReadPtrAt(size_t buffer_offset, T** ptr,
+                               size_t ptr_num) const {
+    size_t size = sizeof(T) * ptr_num;
+    base::CheckedNumeric<size_t> len = buffer_offset;
+    size_t size_needed;
+    if (!(len + size).AssignIfValid(&size_needed)) return false;
+    if (size_needed > buffer_len_) {
+      return false;
+    }
```
Comment (TomTaehoonKim) [inline tachyon/base/buffer/read_only_buffer.h:146]:
```suggestion
    if (size_needed > buffer_len_) return false;
```
This is optional and you may resolve this comment if you think it's not necessary.

Comment (chokobole) [reply tachyon/base/buffer/read_only_buffer.h:146]: I agreed that this looks better, but if you see `tachyon/baes/buffer/read_only_buffer.cc`, there are 2 more places that use this pattern, so I'll resolve Thanks :)

File: tachyon/base/maybe_owned.h [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    MaybeOwnedTraits<T>::ResetPtr(&other.ptr_);
+  }
+  constexpr MaybeOwned& operator=(MaybeOwned&& other) {
+    value_ = std::move(other.value_);
+    owned_ = std::exchange(other.owned_, false);
+    UpdatePtr(other);
+    MaybeOwnedTraits<T>::ResetPtr(&other.ptr_);
+    return *this;
+  }
+  // NOTE(chokobole): |explicit| keyword is explicitly removed for convenience.
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/base/maybe_owned.h:42]: Yes because it's a copy constructor.

File: tachyon/zk/r1cs/groth16/verifying_key.h [hunk 3]
```diff
@@ -110,6 +112,23 @@ class VerifyingKey : public Key {
...
 
+  std::string ToString() const {
+    return absl::Substitute(
+        "{alpha_g1: $0, beta_g2: $1, gamma_g2: $2, delta_g2: $3, l_g1_query: "
+        "$4}",
+        alpha_g1_.ToString(), beta_g2_.ToString(), gamma_g2_.ToString(),
+        delta_g2_.ToString(), base::ContainerToString(l_g1_query_));
+  }
+
+  std::string HexToString(bool pad_zero = false) const {
```
Comment (batzor) [inline tachyon/zk/r1cs/groth16/verifying_key.h:135]:
```suggestion
  std::string ToHexString(bool pad_zero = false) const {
```

File: tachyon/zk/r1cs/groth16/prepared_verifying_key.h [hunk 4]
```diff
@@ -40,6 +41,24 @@ class PreparedVerifyingKey {
...
 
+  std::string ToString() const {
+    return absl::Substitute(
+        "{verifying_key: $0, alpha_g1_beta_g2: $1, delta_neg_g2: $2, "
+        "gamma_neg_g2: $3}",
+        verifying_key_.ToString(), alpha_g1_beta_g2_.ToString(),
+        delta_neg_g2_.ToString(), gamma_neg_g2_.ToString());
+  }
+
+  std::string HexToString(bool pad_zero = false) const {
```
Comment (batzor) [inline tachyon/zk/r1cs/groth16/prepared_verifying_key.h:51]:
```suggestion
  std::string ToHexString(bool pad_zero = false) const {
```

File: tachyon/zk/r1cs/groth16/proving_key.h [hunk 5]
```diff
@@ -157,6 +158,31 @@ class ProvingKey : public Key {
...
+        "b_g1_query: $4, b_g2_query: $5, h_g1_query: $6, l_g1_query}",
+        verifying_key_.ToString(), beta_g1_.ToString(), delta_g1_.ToString(),
+        base::ContainerToString(a_g1_query_),
+        base::ContainerToString(b_g1_query_),
+        base::ContainerToString(b_g2_query_),
+        base::ContainerToString(h_g1_query_),
+        base::ContainerToString(l_g1_query_));
+  }
+
+  std::string HexToString(bool pad_zero = false) const {
```
Comment (batzor) [inline tachyon/zk/r1cs/groth16/proving_key.h:193]:
```suggestion
  std::string ToHexString(bool pad_zero = false) const {
```

File: tachyon/zk/r1cs/groth16/proving_key.h [hunk 6]
```diff
@@ -157,6 +158,31 @@ class ProvingKey : public Key {
...
+    return absl::Substitute(
+        "{verifying_key: $0, beta_g1: $1, delta_g1: $2, a_1_query: $3, "
+        "b_g1_query: $4, b_g2_query: $5, h_g1_query: $6, l_g1_query}",
```
Comment (batzor) [inline tachyon/zk/r1cs/groth16/proving_key.h:196]:
```suggestion
        "b_g1_query: $4, b_g2_query: $5, h_g1_query: $6, l_g1_query: $7}",

```

File: vendors/circom/circomlib/wtns/wtns_unittest.cc [hunk 7]
```diff
@@ -31,15 +31,14 @@ TEST(WtnsTest, Parse) {
...
-      F(4),
-      F(5),
-      F(12),
-  }};
-  EXPECT_EQ(wtns->ToV2()->data, expected_data);
+  std::vector<F> expected_data{
+      F::FromMontgomery(F(1).ToBigInt()), F::FromMontgomery(F(60).ToBigInt()),
+      F::FromMontgomery(F(3).ToBigInt()), F::FromMontgomery(F(4).ToBigInt()),
+      F::FromMontgomery(F(5).ToBigInt()), F::FromMontgomery(F(12).ToBigInt()),
+  };
```
Comment (batzor) [inline vendors/circom/circomlib/wtns/wtns_unittest.cc:38]:
```suggestion
  std::vector<F> expected_data{
      F(1), F(60), F(3), F(4), F(5), F(12),
  };
```

Comment (chokobole) [reply vendors/circom/circomlib/wtns/wtns_unittest.cc:38]: Witnesses are stored in unreduced montgomery form, this transformation is needed!

File: vendors/circom/README.md [hunk 8]
```diff
@@ -73,4 +73,7 @@ Optional arguments:
 --curve             The curve type among ('bn254', bls12_381'), by default 'bn254'
 --no_zk             Create proof without zk. By default zk is enabled. Use this flag in case you want to compare the proof with rapidsnark.
 --verify            Verify the proof. By default verify is disabled. Use this flag to verify the proof with the public inputs.
+-n, --num_runs      The number of times to run the proof generation
```
Comment (TomTaehoonKim) [inline vendors/circom/README.md:76]: Can you make a separate commit or add to commit message about this change?

**Review by batzor — APPROVED** [review]
LGTM

File: vendors/circom/circomlib/zkey/coefficient.h [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+#include <string>
+
+#include "absl/strings/substitute.h"
+
+namespace tachyon::circom {
+
+#pragma pack(push, 1)
+template <typename F>
+struct Coefficient {
+  // if |matrix| is 0, then |a|, otherwise |b| matrix.
```
Comment (ashjeong) [inline vendors/circom/circomlib/zkey/coefficient.h:15]: Could you elaborate on what `|a|` and `|b|` are?

File: vendors/circom/circomlib/zkey/zkey.h [hunk 10]
```diff
@@ -12,14 +12,16 @@
...
+#include "circomlib/zkey/coefficient.h"
 #include "circomlib/zkey/proving_key.h"
 #include "tachyon/base/auto_reset.h"
 #include "tachyon/base/buffer/endian_auto_reset.h"
 #include "tachyon/base/buffer/read_only_buffer.h"
 #include "tachyon/base/files/file_util.h"
 #include "tachyon/base/logging.h"
+#include "tachyon/base/openmp_util.h"
 #include "tachyon/base/strings/string_util.h"
+#include "tachyon/zk/r1cs/constraint_system/constraint_matrices.h"
```
Comment (GideokKim) [inline vendors/circom/circomlib/zkey/zkey.h:24]: I think it would be better to add a dependency to `BUILD.bazel` as well.

File: benchmark/msm/README.md [hunk 11]
```diff
@@ -110,6 +114,10 @@ bazel run -c opt --//:has_openmp --//:has_rtti --//:has_matplotlib //benchmark/m
 
 ### Non-uniform points
 
+```
```
Comment (GideokKim) [inline benchmark/msm/README.md:117]:
```suggestion
```shell
```

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/base/maybe_owned.h [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+  //
+  // class Storage {
+  //  public:
+  //   Storage(const std::string& data): data_(data) {}
+  //   Storage(std::string&& data): data_(std::move(data)) {}
+  //  private:
+  //   MaybeOwned<std::string> data_;
+  // };
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/base/maybe_owned.h:50]: It's a code example to explain my intension. Please see similar examples in `tachyon/base/scoped_generic.h`!

File: vendors/circom/circomlib/zkey/coefficient.h (hunk 9 above)
Comment (chokobole) [reply vendors/circom/circomlib/zkey/coefficient.h:15]: `a`, `b` are the ones that is used in a R1CS `A * z (Had) B * z = C * z`

Comment (ashjeong) [reply vendors/circom/circomlib/zkey/coefficient.h:15]: Ah I see. Would it be possible to include more comments for these `unit32_t` members? I think it would be helpful, especially since `|a|` and `|b|` weren't mentioned anywhere else in this file so I was confused.

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by GideokKim — APPROVED** [review]
LGTM

File: vendors/circom/circomlib/zkey/coefficient.h [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+//
+// - [aᵢ,₀, ...,  aᵢ,ₘ₋₁] * [z₀, ..., zₘ₋₁]
+// - [bᵢ,₀, ...,  bᵢ,ₘ₋₁] * [z₀, ..., zₘ₋₁]
+// - [cᵢ,₀, ...,  cᵢ,ₘ₋₁] * [z₀, ..., zₘ₋₁]
+//
+// where i is the index of the constraints (0 ≤ i < n),
+// m is the number of QAP variables, and n is the number of constraints.
+//
+// The last one is computed if we know the first two ones.
+// Therefore, |Coefficient| represents the first two ones.
```
Comment (ashjeong) [inline vendors/circom/circomlib/zkey/coefficient.h:24]:
```suggestion
// The last constraint is computed if we know the first two constraints.
// Therefore, |Coefficient| represents the first two constraints.

File: vendors/circom/circomlib/zkey/coefficient.h [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+template <typename F>
+struct Coefficient {
+  // if |matrix| is 0, then this constraint is for A matrix, otherwise B matrix.
```
Comment (ashjeong) [inline vendors/circom/circomlib/zkey/coefficient.h:27]: To follow the standard of the other variable comments:
```suggestion
  // A value denoting the matrix this constraint is for. If 0, this constraint is for matrix. Else, this constraint is for matrix B.

File: vendors/circom/circomlib/zkey/coefficient.h [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+  uint32_t matrix;
+  // The index of the constraint, (0 ≤ i < n).
+  uint32_t constraint;
+  // The index of the QAP variables, (0 ≤ j < m).
+  uint32_t signal;
+  // The values of the coefficient; if the |matrix| is 0, then this points to
+  // the a[i][j], otherwise b[i][j].
```
Comment (ashjeong) [inline vendors/circom/circomlib/zkey/coefficient.h:34]:
```suggestion
  // the a[i][j]. Otherwise, this points to b[i][j].

**Review by ashjeong — CHANGES_REQUESTED** [review]

**Review by ashjeong — APPROVED** [review]
LGTM
